### PR TITLE
feat(pin_cite): Add pin_cite span helper

### PR DESCRIPTION
**Pull request by flooie**
Add a span pincite helper for full case citations

This should make it easier to annotate pincites

### CODE REVIEW

File: eyecite/models.py [hunk 1]
```diff
@@ -73,6 +73,7 @@ class CitationBase:
     span_end: Optional[int] = None
     full_span_start: Optional[int] = None
     full_span_end: Optional[int] = None
+    pin_cite_start: Optional[int] = None
```
Comment (grossir) [inline eyecite/models.py:76]: I would put this inside `citation.metadata`, just to not clutter the `citation` interface
Also, would change the name to `pincite_span_start`, to keep the naming conventions

File: tests/test_AnnotateTest.py [hunk 2]
```diff
@@ -240,3 +240,49 @@ def test_long_diff(self):
...
+            # capture supra pin cite span in html
+            (
+                "Bell Atlantic Corp. </em>v. <em>Twombly, </em>550 U. S. 544 (2007), which discussed... apellate court’s core competency <em>Twombly, </em>550 U. S., at 557. Evaluating... In <em>Twombly</em>, supra, at 553-554, the ",
+                ["550 U. S. 544", "550 U. S., at 557", "supra, at 553-554"],
+            ),
+            # can we wrap the full pincite that appears before
+            # https://www.courtlistener.com/opinion/8524158/in-re-cahill/
+            (
+                "principal residence.” Nobelman at 332, 113 S.Ct. 2106",
+                ["Nobelman at 332, 113 S.Ct. 2106"],
```
Comment (grossir) [inline tests/test_AnnotateTest.py:275]: in this case, it's the same as the full_span when it shouldn't; check upper comments

File: eyecite/models.py [hunk 3]
```diff
@@ -184,15 +185,23 @@ def span(self):
...
+        start = min(
+            filter(
+                None, [self.pin_cite_start, self.span_start, self.token.start]
             ),
-            self.pin_cite_end if self.pin_cite_end is not None else self.token.end,
+            default=self.token.start,
+        )
+        end = max(
+            filter(None, [self.pin_cite_end, self.token.end, self.span_end]),
+            default=None,
```
Comment (grossir) [inline eyecite/models.py:202]: shouldn't this default to the span end or token end?; potentially dangerous to return None, may return the whole string
```python
In [2]: "this is a string"[1:None]
Out[2]: 'his is a string'
```

File: eyecite/models.py [hunk 4]
```diff
@@ -184,15 +185,23 @@ def span(self):
         )
 
     def span_with_pincite(self):
```
Comment (grossir) [inline eyecite/models.py:187]: missing type annotation

File: eyecite/helpers.py [hunk 5]
```diff
@@ -185,6 +185,11 @@ def add_pre_citation(citation: FullCaseCitation, words: Tokens) -> None:
     if not m:
         return
 
+    if m["pin_cite"]:
+        # if a pin cite occurs before the citation mark it down
+        start, end = m.span()
```
[the remark this answers is not in the document]
Comment (flooie) [reply eyecite/helpers.py:190]: yes - I realize that - but made the calculated choice because of hte uniqueness of it and how awkward it would look otherwise.

File: eyecite/models.py (hunk 4 above)
Comment (flooie) [reply eyecite/models.py:187]: sure thanks

File: tests/test_AnnotateTest.py (hunk 2 above)
Comment (flooie) [reply tests/test_AnnotateTest.py:275]: yes, this was a conscious choice here because of the edge-case-y-ness of this example

File: eyecite/models.py (hunk 1 above)
Comment (flooie) [reply eyecite/models.py:76]: split the different pin_cite_span_start

File: eyecite/helpers.py [hunk 6]
```diff
@@ -185,6 +185,11 @@ def add_pre_citation(citation: FullCaseCitation, words: Tokens) -> None:
...
+        citation.pin_cite_start = citation.span()[0] - (end - start)
```
[the remark this answers is not in the document]
Comment (flooie) [reply eyecite/helpers.py:191]: sure - okay,

**Review by grossir — COMMENTED** [review]
Looks good; you have to solve the lint and it's ready to go

**Comment by flooie** [comment]
@grossir back to you - I moved the pincite span to metadata and updated tests accordingly.
